src/plugincollection.js

```diff
@@ -0,0 +1,144 @@
+			// done.
+			return Promise.all( plugins.map( pluginPromise ) );
+
+			// Returns a promise that will load the plugin and add it to the collection before resolving.
+			function pluginPromise( plugin ) {
+				return new Promise( function( resolve, reject ) {
+					// Do nothing if the plugin is already loaded (or if is being loaded right now).
+					if ( that._names[ plugin ] || loading[ plugin ] ) {
+						resolve();
+
+						return;
+					}
+
+					CKEDITOR.require( [ 'plugin!' + plugin ],
+						// Success callback.
+						function( LoadedPlugin ) {
+							var loadedPlugin = new LoadedPlugin( that._editor );
+							loadedPlugin.name = plugin;
+							loadedPlugin.path = CKEDITOR.getPluginPath( plugin );
+							loadedPlugin.deps = getPluginDeps( plugin );
+
+							loading[ plugin ] = true;
+
+							// Resolve with a promise that resolves once all dependencies are loaded.
+							resolve(
+								Promise.all( loadedPlugin.deps.map( pluginPromise ) )
+									.then( function() {
+										// Once dependencies are loaded, add the new instance of the loaded plugin to
+										// the collection. This guarantees that dependecies come first in the collection.
+										that.add( loadedPlugin );
+									} )
+							);
+						},
+						// Error callback.
+						function() {
+							var err = new Error( 'It was not possible to load the "' + plugin + '" plugin.' );
+							err.name = 'CKEditor Error';
+							reject( err );
+						}
+					);
+				} );
+			}
+
+			function getPluginDeps( name ) {
+				// Get the list of AMD modules that the plugin depends on.
+				var deps = CKEDITOR._dependencies[ 'plugin!' + name ] || [];
+
+				deps = deps
+					// Pick only dependencies that are other plugins.
+					.filter( function( dep ) {
+						return dep.indexOf( 'plugin!' ) === 0;
+					} )
+					// Remove the 'plugin!' prefix.
+					.map( function( dep ) {
+						return dep.substr( 7 );
+					} );
+
+				return deps;
+			}
+		},
+
+		/**
+		 * Adds a plugin to the collection.
+		 *
+		 * The `name` property must be set to the plugin object before passing it to this function. Adding plugins
+		 * with the same name has no effect and silently fails.
+		 *
+		 * @param {Plugin} plugin The plugin to be added.
+		 */
+		add: function( plugin ) {
+			// Do nothing if the plugin is already loaded.
+			if ( this._names[ plugin.name ] ) {
+				return;
+			}
+
+			// Save a pointer to the plugin by its name.
+			this._names[ plugin.name ] = plugin;
+
+			// Call the original implementation.
+			Collection.prototype.add.apply( this, arguments );
+		},
+
+		/**
+		 * Gets a plugin from the collection.
+		 *
+		 * @param {String} name The plugin name.
+		 * @returns {Plugin} The requested plugin, if available in the collection.
+		 */
+		get: function( name ) {
+			if ( typeof name != 'string' ) {
+				return Collection.prototype.get.apply( this, arguments );
```

Comment:
You mean, the ability to retrieve a plugin by name? Yes this is very important when one wants to interacts with the plugin instance in the editor directly, for example calling methods of that plugin.